service/history/workflow/mutable_state_impl.go

```diff
@@ -5137,14 +5133,112 @@ func (ms *MutableStateImpl) prepareCloseTransaction(
 		}
 	}
 
+	return nil
+}
+
+func (ms *MutableStateImpl) closeTransactionHandleSpeculativeWorkflowTask(
+	transactionPolicy TransactionPolicy,
+) error {
+	if transactionPolicy == TransactionPolicyPassive ||
+		!ms.IsWorkflowExecutionRunning() {
+		return nil
+	}
+
+	// It is important to convert speculative WT to normal before prepareEventsAndReplicationTasks,
+	// because prepareEventsAndReplicationTasks will move internal buffered events to the history,
+	// and WT related events (WTScheduled, in particular) need to go first.
+	return ms.workflowTaskManager.convertSpeculativeWorkflowTaskToNormal()
+}
+
+func (ms *MutableStateImpl) closeTransactionUpdateTransitionHistory(
+	transactionPolicy TransactionPolicy,
+	workflowEventsSeq []*persistence.WorkflowEvents,
+) error {
+
+	if len(workflowEventsSeq) > 0 {
+		lastEvents := workflowEventsSeq[len(workflowEventsSeq)-1].Events
+		lastEvent := lastEvents[len(lastEvents)-1]
+		if err := ms.updateWithLastWriteEvent(
+			lastEvent,
+			transactionPolicy,
+		); err != nil {
+			return err
+		}
+	}
+
+	if transactionPolicy != TransactionPolicyActive {
+		return nil
+	}
+
+	if !ms.config.EnableMutableStateTransitionHistory() {
```

Comment:
Do we still need this feature flag? I think we're good with your design.

Reply:
I'd like to play safe here, and keep it disabled until we've test replication E2E.

Reply:
actually I need to keep it disabled for xdc or buffered events can't be handled correctly upon failover.
Check [here](https://github.com/temporalio/temporal/pull/5872/files#diff-8b569e7d448b915cfc54cba06dfa7dfa8d639218eab08286169ede455af3dd10R5536) for more details.